chainbase/src/main/java/org/tron/core/service/RewardCalService.java

```diff
@@ -0,0 +1,188 @@
+      }
+    }
+    es.submit(this::startRewardCal);
+  }
+
+  public void calRewardForTest() throws IOException {
+    newRewardCalStartCycle = propertiesStore.getNewRewardAlgorithmEffectiveCycle();
+    isDoneKey = ByteArray.fromLong(newRewardCalStartCycle);
+    if (rewardCacheStore.has(isDoneKey)) {
+      logger.info("RewardCalService is already done");
+      return;
+    }
+    accountIterator = (DBIterator) accountStore.getDb().iterator();
+    try (DBIterator iterator = rewardCacheStore.iterator()) {
+      iterator.seekToLast();
+      if (iterator.hasNext()) {
+        byte[] key  = iterator.next().getKey();
+        System.arraycopy(key, 0, lastAccount, 0, ADDRESS_SIZE);
+      }
+    }
+    startRewardCal();
+  }
+
+
+  private void startRewardCal() {
+    if (!doing.compareAndSet(false, true)) {
+      logger.info("RewardCalService is doing");
+      return;
+    }
+    logger.info("RewardCalService start from lastAccount: {}", ByteArray.toHexString(lastAccount));
+    ((DBIterator) delegationStore.getDb().iterator()).prefixQueryAfterThat(
+        new byte[]{Constant.ADD_PRE_FIX_BYTE_MAINNET}, lastAccount).forEachRemaining(e -> {
+          try {
+            doRewardCal(e.getKey(), ByteArray.toLong(e.getValue()));
+          } catch (InterruptedException error) {
+            Thread.currentThread().interrupt();
+          }
+        });
+    rewardCacheStore.put(this.isDoneKey, IS_DONE_VALUE);
+    logger.info("RewardCalService is done");
+  }
+
+  private void doRewardCal(byte[] address, long beginCycle) throws InterruptedException {
+    if (beginCycle >= newRewardCalStartCycle) {
+      return;
+    }
+    long endCycle = delegationStore.getEndCycle(address);
+    //skip the last cycle reward
+    if (beginCycle + 1 == endCycle) {
+      beginCycle += 1;
+    }
+    if (beginCycle >= newRewardCalStartCycle) {
+      return;
+    }
+
+    Protocol.Account account = getAccount(address);
+    if (account == null || account.getVotesList().isEmpty()) {
+      return;
+    }
+    Histogram.Timer requestTimer = Metrics.histogramStartTimer(
+        MetricKeys.Histogram.DO_REWARD_CAL_DELAY,
+        (newRewardCalStartCycle - beginCycle) / 100 + "");
+    long reward = LongStream.range(beginCycle, newRewardCalStartCycle)
+        .map(i -> computeReward(i, account))
+        .sum();
+    rewardCacheStore.putReward(Bytes.concat(address, Longs.toByteArray(beginCycle)), reward);
+    Metrics.histogramObserve(requestTimer);
+  }
+
+  private Protocol.Account getAccount(byte[] address) {
+    try {
+      accountIterator.seek(address);
+      if (accountIterator.hasNext()) {
+        byte[] account = accountIterator.next().getValue();
+        return Protocol.Account.parseFrom(account);
+      }
+    } catch (InvalidProtocolBufferException e) {
+      logger.error("getAccount error: {}", e.getMessage());
+    }
+    return null;
+  }
+
+  long computeReward(long cycle, Protocol.Account account) {
```

Comment:
Why rewrite this method instead of calling the original method?

Reply:
Avoid circular dependencies.

Reply:
done